eventstore/query_json.go

```diff
@@ -0,0 +1,232 @@
+	dsq := dsquery.Query{
+		Prefix: t.model.dsKey.String(),
+	}
+	dsr, err := t.model.store.datastore.Query(dsq)
+	if err != nil {
+		return nil, fmt.Errorf("error when internal query: %v", err)
+	}
+
+	var values []marshaledValue
+	for {
+		res, ok := dsr.NextSync()
+		if !ok {
+			break
+		}
+		val := make(map[string]interface{})
+		if err := json.Unmarshal(res.Value, &val); err != nil {
+			return nil, fmt.Errorf("error when unmarshaling query result: %v", err)
+		}
+		ok, err = q.matchJSON(val)
+		if err != nil {
+			return nil, fmt.Errorf("error when matching entry with query: %v", err)
+		}
+		if ok {
+			values = append(values, marshaledValue{value: val, rawJson: res.Value})
+		}
+	}
+
+	if q.Sort.FieldPath != "" {
+		var wrongField, cantCompare bool
+		sort.Slice(values, func(i, j int) bool {
+			fieldI, err := traverseFieldPathMap(values[i].value, q.Sort.FieldPath)
+			if err != nil {
+				wrongField = true
+				return false
+			}
+			fieldJ, err := traverseFieldPathMap(values[j].value, q.Sort.FieldPath)
+			if err != nil {
+				wrongField = true
+				return false
+			}
+			res, err := compare(fieldI.Interface(), fieldJ.Interface())
+			if err != nil {
+				cantCompare = true
+				return false
+			}
+			if q.Sort.Desc {
+				res *= -1
+			}
+			return res < 0
+		})
+		if wrongField {
+			return nil, ErrInvalidSortingField
+		}
+		if cantCompare {
+			panic("can't compare while sorting")
+		}
+	}
+
+	res := make([]string, len(values))
+	for i := range values {
+		res[i] = string(values[i].rawJson)
+	}
+
+	return res, nil
+}
+
+func (q *JSONQuery) matchJSON(v map[string]interface{}) (bool, error) {
+	if q == nil {
+		panic("query can't be nil")
+	}
+
+	andOk := true
+	for _, c := range q.Ands {
+		fieldRes, err := traverseFieldPathMap(v, c.FieldPath)
+		if err != nil {
+			return false, err
+		}
+		ok, err := c.match(fieldRes)
+		if err != nil {
+			return false, err
+		}
+		andOk = andOk && ok
+		if !andOk {
+			break
+		}
+	}
+	if andOk {
+		return true, nil
+	}
+
+	for _, q := range q.Ors {
+		ok, err := q.matchJSON(v)
+		if err != nil {
+			return false, err
+		}
+		if ok {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func compareJsonValue(value interface{}, critVal JSONValue) (int, error) {
+	if critVal.String != nil {
+		s, ok := value.(string)
+		if !ok {
+			return 0, &errTypeMismatch{value, critVal}
+		}
+		return strings.Compare(s, *critVal.String), nil
+	}
+	if critVal.Bool != nil {
+		b, ok := value.(bool)
+		if !ok {
+			return 0, &errTypeMismatch{value, critVal}
+		}
+		if *critVal.Bool == b {
+			return 0, nil
+		}
+		return -1, nil
+	}
+	if critVal.Float != nil {
+		f, ok := value.(float64)
+		if !ok {
+			return 0, &errTypeMismatch{value, critVal}
+		}
+		if f == *critVal.Float {
+			return 0, nil
+		}
+		if f < *critVal.Float {
+			return -1, nil
+		}
+		return 1, nil
+	}
+	log.Fatalf("no underlying value for json criterion was provided")
+	return 0, nil
+}
```

Comment:
This is somewhat analogous with the [`compare()` in `criterion.go`](https://github.com/textileio/go-textile-threads/blob/master/eventstore/criterion.go#L186). As mentioned in #131, comparing both functions makes clear the point of the limitations of JSONMode queries vs Go typed queries.